src/compiler/binder.ts

```diff
@@ -130,7 +130,10 @@ namespace ts {
                 symbol.members = {};
             }
 
-            if (symbolFlags & SymbolFlags.Value && !symbol.valueDeclaration) {
+            if (symbolFlags & SymbolFlags.Value &&
+                (!symbol.valueDeclaration ||
+                 (symbol.valueDeclaration.kind === SyntaxKind.ModuleDeclaration && node.kind !== SyntaxKind.ModuleDeclaration))) {

```

Comment:
That's a nice rewording. I took the change except for inverting the predicate -- I can't do that without wrapping the whole thing in a `!` or having an empty `if` body and an extra `else`.